File: maximum_wealth/maximum_wealth.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-02-16 21:53:30
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-02-16 22:08:17
 */

#include <cassert>
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

int maximum_wealth(vector<vector<int>>& accounts) {
  int res = INT_MIN;
  int count = 0;
  for (vector<int>& account : accounts) {
    for (int num : account) {
      count += num;
    }
    res = max(res, count);
    count = 0;
  }

  return res;
}

int main(void) {
  vector<vector<int>> account;
  account = {{1, 2, 3}, {3, 2, 1}};
  assert(6 == maximum_wealth(account));
  account = {{1, 5}, {7, 3}, {3, 5}};
  assert(10 == maximum_wealth(account));
  account = {{2, 8, 7}, {7, 1, 3}, {1, 9, 5}};
  assert(17 == maximum_wealth(account));

  return 0;
}

File: rust/maximum_wealth/maximum_wealth.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-02-16 21:53:28
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-02-16 22:15:20
 */

fn maximum_wealth(accounts: Vec<Vec<i32>>) -> i32 {
    accounts
        .into_iter()
        .map(|v| v.into_iter().sum())
        .max()
        .unwrap()
}

fn main() {
    let mut accounts: Vec<Vec<i32>>;
    accounts = vec![vec![1, 2, 3], vec![3, 2, 1]];
    assert_eq!(6, maximum_wealth(accounts));
    accounts = vec![vec![1, 5], vec![7, 3], vec![3, 5]];
    assert_eq!(10, maximum_wealth(accounts));
    accounts = vec![vec![2, 8, 7], vec![7, 1, 3], vec![1, 9, 5]];
    assert_eq!(17, maximum_wealth(accounts));
}
